crates/turbo-tasks/src/trait_ref.rs

```diff
@@ -0,0 +1,64 @@
+use std::{fmt::Debug, future::Future, marker::PhantomData, sync::Arc};
+
+use anyhow::Result;
+use serde::{Deserialize, Serialize};
+
+use crate::{manager::find_cell_by_type, RawVc, SharedReference, ValueTraitVc};
+
+/// Similar to a [`ReadRef<T>`], but contains a value trait object instead. The
+/// only way to interact with a `TraitRef<T>` is by passing it around or turning
+/// it back into a value trait vc by calling [`ReadRef::cell`].
+///
+/// Internally it stores a reference counted reference to a value on the heap.
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
+pub struct TraitRef<T>
+where
+    T: ?Sized,
+{
+    shared_reference: SharedReference,
+    _t: PhantomData<Arc<T>>,
+}
+
+impl<T> TraitRef<T> {
+    /// # Safety
+    ///
+    /// The caller must ensure that:
+    /// * the shared reference points to a vc that implements the value trait T;
+    /// * the shared reference is typed.
+    pub(crate) unsafe fn new(shared_reference: SharedReference) -> Self {
```

Comment:
```suggestion
    pub(crate) fn new(shared_reference: SharedReference) -> Self {
```

Reply:
I marked all of these as unsafe as there are some assumptions the caller must verify in order to use these methods soundly. It's not related to memory safety, but I think it's still a useful tool to ensure we're not calling into any sensitive methods by accident.

Reply:
I can remove all these `unsafe` if we don't want to mix the two.